Source/Core/Core/Movie.cpp

```diff
@@ -854,9 +854,9 @@ void RecordInput(GCPadStatus* PadStatus, int controllerID)
 
   CheckPadStatus(PadStatus, controllerID);
 
-  EnsureTmpInputSize((size_t)(s_currentByte + 8));
-  memcpy(&(tmpInput[s_currentByte]), &s_padState, 8);
-  s_currentByte += 8;
+  EnsureTmpInputSize((size_t)(s_currentByte + sizeof(ControllerState)));
+  *reinterpret_cast<ControllerState*>(&tmpInput[s_currentByte]) = s_padState;
```

Comment:
Here's the whole overview, complete with ~pedantic~ standard references/(specifically, the C++14 and C11 standards) in case you want to check out anything the quotes link to:

Yeah, this would be undefined behavior, because they're unrelated types. `memcpy` is the only valid way to copy the bits between them.

Specifically the standard states (In section 3.10.10 aka [basic.lval]/10 of the C++14 standard):

> If a program attempts to access the stored value of an object through a glvalue of other than one of the following types. the behavior is undefined:
> - the dynamic type of the object,
> - a cv-qualified version of the dynamic type of the object,
> - a type similar (as defined in 4.4) to the dynamic type of the object,
> - a type that is the signed or unsigned type corresponding to the dynamic type of the object,
> - a type that is the signed or unsigned type corresponding to a cv-qualified version of the dynamic type of the object,
> - an aggregate or union type that includes one of the aforementioned types among its elements or nonstatic data members (including, recursively, an element or non-static data member of a subaggregate or contained union),
> - a type that is a (possibly cv-qualified) base class type of the dynamic type of the object,
> - a char or unsigned char type.

As for why memcpy is valid, it's defined as internally representing both its source and destination arguments as character arrays, which satisfies one of the listed exceptions (the last one).

The standardized definition being (italics added by me for emphasis) (section 7.24.2.1 in the C11 standard):

> The memcpy function copies n *characters* from the object pointed to by s2 into the object pointed to by s1. If copying takes place between objects that overlap, the behavior is undefined.

and C++ defines character sequences as (section 17.5.2.1.4 in the C++14 standard):

> A character sequence is an array object (8.3.4) A that can be declared as `T A[N]`, where `T` is any of the types `char`, `unsigned char`, or `signed char` (3.9.1), optionally qualified by any combination of `const` or `volatile`.

So memcpy is solid when it comes to satisfying the last bullet point.

As an aside, since this is a similar common misconception: Using unions as an alternative way to type pun without `reinterpret_cast` like so:

```cpp
union Converter
{
    s32 integral;
    float real;
};

Converter c;
c.f = 0.5f;

// Get integral equivalent
s32 value = c.i;
```

is also undefined behavior in C++ (however, C has allowed this since C99).